### PR TITLE
fix segfault when no versions for a pacakge is found

### LINKED ISSUE #2724: null Package pointer
### System information

- **dub version**: DUB version 1.35.0, built on Dec  2 2023
- **OS Platform and distribution**:Linux
- **compiler version** DMD64 D Compiler v2.106.0

### Bug Description

The `void collectDependenciesRec(Package pack, int depth = 0)` function get's a null package reference:

```
core.exception.AssertError@source/dub/project.d(475): pack is null
----------------
??:? _d_assert_msg [0x108a878]
source/dub/project.d:475 void dub.project.Project.reinit().collectDependenciesRec(dub.package_.Package, int) [0x1036b55]
source/dub/project.d:558 void dub.project.Project.reinit().collectDependenciesRec(dub.package_.Package, int) [0x1037824]
source/dub/project.d:565 void dub.project.Project.reinit() [0x10369cc]
source/dub/project.d:90 dub.project.Project dub.project.Project.__ctor(dub.packagemanager.PackageManager, dub.package_.Package) [0x1033670]
source/dub/project.d:71 dub.project.Project dub.project.Project.__ctor(dub.packagemanager.PackageManager, dub.internal.vibecompat.inet.path.NativePath) [0x103352d]
source/dub/dub.d:458 void dub.dub.Dub.loadPackage(dub.internal.vibecompat.inet.path.NativePath) [0xdf85cd]
source/dub/dub.d:452 void dub.dub.Dub.loadPackage() [0xdf8594]
source/dub/commandline.d:887 bool dub.commandline.Command.loadCwdPackage(dub.dub.Dub, bool) [0xc18fa5]
source/dub/commandline.d:1354 bool dub.commandline.PackageBuildCommand.loadSpecificPackage(dub.dub.Dub, immutable(char)[], immutable(char)[]) [0xc1bd4b]
source/dub/commandline.d:1306 void dub.commandline.PackageBuildCommand.setupPackage(dub.dub.Dub, immutable(char)[], immutable(char)[], immutable(char)[]) [0xc1b80b]
source/dub/commandline.d:1283 void dub.commandline.PackageBuildCommand.setupVersionPackage(dub.dub.Dub, immutable(char)[], immutable(char)[]) [0xc1b403]
source/dub/commandline.d:1436 int dub.commandline.GenerateCommand.execute(dub.dub.Dub, immutable(char)[][], immutable(char)[][]) [0xc1c59a]
source/dub/commandline.d:1514 int dub.commandline.BuildCommand.execute(dub.dub.Dub, immutable(char)[][], immutable(char)[][]) [0xc1cdd6]
source/dub/commandline.d:1604 int dub.commandline.RunCommand.execute(dub.dub.Dub, immutable(char)[][], immutable(char)[][]) [0xc1d51e]
source/dub/commandline.d:509 int dub.commandline.runDubCommandLine(immutable(char)[][]) [0xc1738b]
source/app.d:26 _Dmain [0xbf4303]
```

### PR DESCRIPTION
**Pull request by gedaiu**
fix #2724

### CODE REVIEW

File: source/dub/project.d [hunk 1]
```diff
@@ -545,7 +545,7 @@ class Project {
 					}
 				}
 
-				if (!m_dependencies.canFind(p)) {
+				if (!m_dependencies.canFind(p) && p !is null) {
```
Comment (rikkimax) [inline source/dub/project.d:548]:
```suggestion
				if (p !is null && !m_dependencies.canFind(p)) {
```

No need to do the canFind which will be ``O(1)`` if its null.

**Review by rikkimax — CHANGES_REQUESTED** [review]

**Comment by Geod24** [comment]
Under what circumstances does the bug trigger ? Can you add a test ?

**Comment by gedaiu** [comment]
Hey @Geod24, it happens with one of my weirdly configured projects. Unfortunately, I can't reproduce it with a unit test without doing some refactoring in that function.

**Comment by Geod24** [comment]
@gedaiu : Could you share the package recipe (and perhaps selections) of the involved packages ? I might be able to write the UT for you.

**Comment by Geod24** [comment]
Actually there is a very simple test here: https://github.com/dlang/dub/issues/2696
However I was looking at this code and there are more issues with it (namely subpackages would also trigger segv).

**Comment by Geod24** [comment]
Submitted a more complete fix with tests in #2762 . Thanks for your PR though @gedaiu , because it led me to find other bugs!

**Pull request closed without merging**